### ISSUE TITLE
[ENHANCEMENT] Allow upgrading from informal build version to formal release or RC versions

### ISSUE THREAD

**Issue by starbops** [harvester/harvester#5203]
**Is your enhancement related to a problem? Please describe.**

A minimum upgradeable version requirement is set for each RC, as well as a formal release version of Harvester. That implies Harvester will enforce the version checking for the current version's upgrade eligibility.

However, users might have an informal build version of Harvester deployment and want to upgrade it to an RC or formal release version. The minimum upgradability check will then reject the upgrade request. A typical case is that Harvester QA engineers must verify upgrade-related features/bug fixes/enhancements by upgrading a live cluster in an informal version like `v1.2-head` to an RC version.

**Describe the solution you'd like**

Let users have a chance to annotate the Upgrade object with an annotation like `harvesterhci.io/skip-minimum-upgradability-check: true`. Make the upgrade controller and upgrade script honor the annotation, to skip the version checking procedure during the upgrade.

**Describe alternatives you've considered**

Apply the [workaround](https://github.com/harvester/harvester/issues/3040#issuecomment-1298057127) before starting an upgrade.

**Additional context**

This is raised when QA engineers want to verify the upgrade path `v1.2.1` -> `1.2.2` -> `v1.3.0`. At the time, we don't have any RC tags for `v1.2.2` yet. So the effective upgrade path would then become `v1.2.1` -> `v1.2-head` -> `v1.3.0-rc3`. There will be such a problem upgrading from `v1.2-head` to `v1.3.0-rc3`.

Note: Though such upgrade paths would be allowed after the enhancement is delivered, we still don't recommend users take such actions (also not supported).

**Comment by albinsun** [comment]
Provide test result on `v1.3.0-head` and `master`.


## Environment
* Harvester
  * Version: `v1.3.0-head` and `master` -> `v1.3.0-rc3`
  * Profile: QEMU/KVM, single node.
  * ui-source: `Auto`


## Test Case
1. `v1.3-head` -> `v1.3.0-rc3` :heavy_check_mark: 
   <details>
   ![image](https://github.com/harvester/harvester/assets/2773781/10c26093-feb6-4ca9-905c-242a931d9a42)
   ![image](https://github.com/harvester/harvester/assets/2773781/bd2bd4c9-0ca1-42f6-a25e-1e93f1bd72e9)
   </details>

   
1. `master` -> `v1.3.0-rc3` :x: ${\color{red} \text{Invalid Semantic Version}}$
   * Upgrade log: [hvst-upgrade-9wrcx-upgradelog-archive-2024-02-29T12-10-39Z.zip](https://github.com/harvester/harvester/files/14447676/hvst-upgrade-9wrcx-upgradelog-archive-2024-02-29T12-10-39Z.zip)
   * Support bundle: [supportbundle_104d495b-cc6b-4329-a516-988087ec7036_2024-02-29T12-13-41Z.zip](https://github.com/harvester/harvester/files/14447750/supportbundle_104d495b-cc6b-4329-a516-988087ec7036_2024-02-29T12-13-41Z.zip)

   ![image](https://github.com/harvester/harvester/assets/2773781/7eb82c65-9e68-4411-b713-b73a1272d319)
   ![image](https://github.com/harvester/harvester/assets/2773781/b7fadc2e-0131-4bb0-b841-76c15049f0b0)

**Comment by albinsun** [comment]
Provide test result on `v1.3.0-head` and `master`.
The `master` TC will sutck at `Upgrade System Service` phase due to `The Plan "hvst-upgrade-bk7cw-skip-restart-rancher-system-agent" is invalid: spec.version: Invalid value: "integer": spec.version in body must be of type string: "integer"`

Please help to check, thanks.

## Environment
* Harvester
  * Version: `v1.3.0-head` and `master` -> `v1.3.0-rc3`
  * Profile: QEMU/KVM, single node.
  * ui-source: `Auto`


## Test Case
> [!NOTE]
> Following https://github.com/harvester/harvester/pull/5208#issue-2150394040 to annotate upgrade after triggering.

1. `v1.3-head` -> `v1.3.0-rc3` :heavy_check_mark: 
   <details>

   ![image](https://github.com/harvester/harvester/assets/2773781/10c26093-feb6-4ca9-905c-242a931d9a42)
   ![image](https://github.com/harvester/harvester/assets/2773781/bd2bd4c9-0ca1-42f6-a25e-1e93f1bd72e9)
   </details>

   
1. `master` -> `v1.3.0-rc3` :x: ${\color{red} \text{The Plan "hvst-upgrade-bk7cw-skip-restart-rancher-system-agent" is invalid: spec.version: Invalid value: "integer": spec.version in body must be of type string: "integer"}}$
   * Upgrade Log: [hvst-upgrade-bk7cw-upgradelog-archive-2024-02-29T14-52-05Z.zip](https://github.com/harvester/harvester/files/14449583/hvst-upgrade-bk7cw-upgradelog-archive-2024-02-29T14-52-05Z.zip)
   * Support Bundle: [supportbundle_a5ad8d1d-2ebe-485e-b6dd-d95b61e061e7_2024-02-29T14-52-36Z.zip](https://github.com/harvester/harvester/files/14449687/supportbundle_a5ad8d1d-2ebe-485e-b6dd-d95b61e061e7_2024-02-29T14-52-36Z.zip)

   <details open>

   Stuck at `Upgrading System Service`
   ![image](https://github.com/harvester/harvester/assets/2773781/8a58623e-0737-4109-97ff-2794106108f7)
   
   Upgrade pod log
   ```
     harvester-node-0:~ # kubectl -n harvester-system logs pod/hvst-upgrade-bk7cw-apply-manifests-htnzd
   ...
    Current version: master-55356450-head
    Minimum upgradable version: 1.2.1
    Current version is supported.
    managedchart.management.cattle.io/harvester patched
    managedchart.management.cattle.io/harvester-crd patched
    managedchart.management.cattle.io/rancher-monitoring-crd patched
    managedchart.management.cattle.io/rancher-logging-crd patched
    Error from server (NotFound): managedcharts.management.cattle.io "rancher-monitoring" not found
    Error from server (NotFound): managedcharts.management.cattle.io "rancher-logging" not found
    Creating plan hvst-upgrade-bk7cw-skip-restart-rancher-system-agent to make rancher-system-agent temporarily skip restarting RKE2 server...
    The Plan "hvst-upgrade-bk7cw-skip-restart-rancher-system-agent" is invalid: spec.version: Invalid value: "integer": spec.version in body must be of type string: "integer"
   ```
   ![image](https://github.com/harvester/harvester/assets/2773781/a58e006e-1595-4223-9f4d-91a92cd0d4e2)

   </details>

**Comment by starbops** [comment]
@albinsun I've created another issue #5263 to track the bug. Thanks for making the Harvester upgrade more robust.

**Comment by albinsun** [comment]
Provide test result on `v1.3-head` and `master-71a6ac7e-head`, close as fixed.


## Environment
* Harvester
  * Version: `v1.3-head` and `master-71a6ac7e-head`
  * Profile: QEMU/KVM, single node (8C/16G/500G)
  * ui-source: `Auto`


## Steps 
1. Setup single node harvester cluster
   * <details>
     <summary> Cluster is active :heavy_check_mark: 
     </summary>

     ![image](https://github.com/harvester/harvester/assets/2773781/c9a2f4d1-049b-493f-8f25-3135197849c6)
     </details>
   
1. Trigger upgrade to `v1.3.0-rc3` [^1]
   * Should fail w/ <code>Invalid Semantic Version</code> :heavy_check_mark: 


4. Trigger again and annotate it as informal upgrade [^2]
   * Note that we may hit known RC issue cause upgrade stuck , refer https://github.com/harvester/harvester/issues/5149#issuecomment-1960852535 for workaround.
   * <details>
     <summary> Upgrade should success :heavy_check_mark: 
     </summary>
 
     ![image](https://github.com/harvester/harvester/assets/2773781/14b026ce-539b-40ab-ab47-70eb77272799)
     </details>


[^1]: https://docs.harvesterhci.io/v1.1/upgrade/automatic#prepare-an-air-gapped-upgrade
[^2]: https://github.com/harvester/harvester/pull/5208#issue-2150394040
